#! /usr/bin/env python

PACKAGE='humanoid_models'
import roslib; roslib.load_manifest(PACKAGE)

from math import pi

from driver_base.msg import SensorLevels
from dynamic_reconfigure.parameter_generator import *

gen = ParameterGenerator()
#Name            Type   Reconfiguration level             Description         Default Min Max
######### name
gen.add( "humanoid_name",        str_t,    SensorLevels.RECONFIGURE_RUNNING, "humanoid name", "user1" )
gen.add( "sensor_frame_name",    str_t,    SensorLevels.RECONFIGURE_RUNNING, "tf frame observing this humanoid", "/virtual_humanoid_sensor" )
gen.add( "use_displacement",     bool_t,   SensorLevels.RECONFIGURE_RUNNING, "toggle between axis angles or joint displacement", False )

######### neck
gen.add( "neck_yaw",             double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "neck_pitch",           double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "neck_roll",            double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### torso
gen.add( "torso_yaw",            double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "torso_pitch",          double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "torso_roll",           double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )

######### pelvis; this is actually the angle from the virtual sensor to the pelvis
gen.add( "pelvis_yaw",           double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   180, 0, 360 )
gen.add( "pelvis_pitch",         double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "pelvis_roll",          double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )

######### right_shoulder
gen.add( "right_shoulder_yaw",   double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
#gen.add( "right_shoulder_pitch", double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "right_shoulder_roll",  double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### right_elbow
gen.add( "right_elbow_yaw",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "right_elbow_pitch",    double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
#gen.add( "right_elbow_roll",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### right_wrist
gen.add( "right_wrist_yaw",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "right_wrist_pitch",    double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "right_wrist_roll",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### right_hip
gen.add( "right_hip_yaw",        double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "right_hip_pitch",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "right_hip_roll",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### right_knee
#gen.add( "right_knee_yaw",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "right_knee_pitch",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
#gen.add( "right_knee_roll",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### right_ankle
gen.add( "right_ankle_yaw",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "right_ankle_pitch",    double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "right_ankle_roll",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )

######### left_shoulder
gen.add( "left_shoulder_yaw",    double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
#gen.add( "left_shoulder_pitch",  double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "left_shoulder_roll",   double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### left_elbow
gen.add( "left_elbow_yaw",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "left_elbow_pitch",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
#gen.add( "left_elbow_roll",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### left_wrist
gen.add( "left_wrist_yaw",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "left_wrist_pitch",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "left_wrist_roll",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### left_hip
gen.add( "left_hip_yaw",         double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "left_hip_pitch",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "left_hip_roll",        double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### left_knee
#gen.add( "left_knee_yaw",        double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "left_knee_pitch",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
#gen.add( "left_knee_roll",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )
######### left_ankle
gen.add( "left_ankle_yaw",       double_t, SensorLevels.RECONFIGURE_RUNNING, "joint yaw",   0,  -180, 180 )
gen.add( "left_ankle_pitch",     double_t, SensorLevels.RECONFIGURE_RUNNING, "joint pitch", 0,  -180, 180 )
gen.add( "left_ankle_roll",      double_t, SensorLevels.RECONFIGURE_RUNNING, "joint roll",  0,  -180, 180 )

exit(gen.generate(PACKAGE, "dynamic_reconfigure_node", "VirtualHumanoid"))
